src/Statement/StatementList.php

```diff
@@ -325,4 +325,15 @@ public function filter( StatementFilter $filter ) {
 		return $statementList;
 	}
 
+	/**
+	 * @see http://php.net/manual/en/language.oop5.cloning.php
+	 *
+	 * @since 5.1
+	 */
+	public function __clone() {
+		foreach ( $this->statements as &$statement ) {
```

Comment:
The issue here is not that StatementList is mutable but that Statement is mutable. What you are suggesting requires an immutable Statement. Personally I do not think it's worth our time to go this route. As shown on Phabricator this patch saves 90% memory. Having an immutable Statement will probably save some additional 100 KB. Is that worth it?